### DISCUSSION TITLE
Wrangling dependencies

### DISCUSSION THREAD

**Discussion by lukego** [kaveh808/kons-9#191] [General]
How should we manage our software dependencies for kons-9?

It seems to me that kons-9 developers need to have:

- The contents of this repository.
- sbcl.
- Certain Quicklisp libraries without specific version requirements.
- Certain Quicklisp libraries that must be the up-to-date (e.g. `cl-confidence`.)
- Certain Quicklisp libraries that must specifically be different versions than what Quicklisp includes (e.g. `krma` dependencies.)
- Certain non-Quicklisp libraries that need to be installed separately (e.g. `cl-vulkan`.)
- Native libraries that are simple to install e.g. `glfw`.
- Native libraries that are potentially more complex to install e.g. vulkan shader toolchain.

and on the project level we need to:

- Specify these dependencies so that we know what they are.
- Have an easy default way for people to install them.
- Make life easy for people who want to incorporate kons-9 and dependencies into their own build/install processes.

Is that a reasonable assessment? What are our options? What have people had good experience with?

**Comment by kaveh808** [comment]
This is an important discussion to have.

Ease of installation by users, and therefore encouraging user adoption rates, are very important to me. Ideally, down the road at some point, it will be as simple as a double-clickable package users can download.

In the current version of the system, I rolled the text/font dependencies directly into our source and asdf. Perhaps not the most general or elegant solution, but it made those dependencies transparent to the user.

As kons-9 grows, I want to err towards ease of installation.

**Comment by awolven** [comment]
Yes, we should basically make installers for all platforms which would be
the default goto for noobs which requires just running an executable or the
macOS dmg way.  The rest should be our business.

**Comment by lukego** [comment]
Okay, so we want a simple "click to run" solution for end users. How about wrangling dependencies for our own core development purposes though?

Take `krma` for example. On this branch we diverge significantly from Quicklisp because we need forked-and-extended versions of various libraries. I think this will be a really common situation going forward where we need to have "read/write" access to our dependencies meaning users can't install all of them from Quicklisp mainline.

How do we deal with this?

Submodules is one solution as in https://github.com/awolven/krma/tree/main/submodules. I also stumbled today on a project called [Qlot](https://github.com/fukamachi/qlot#what-qlot-is-trying-to-solve) that solves this problem in a different way with a little spec file saying where each package should be installed from (e.g. quicklisp, or specific branch, or specific commit.) I believe that we could also create a custom Quicklisp distribution that uses the standard client but installs the versions that we need (see Ultralisp for an example of different Lisp module distributions distributed using the standard Quicklisp client.)

Thoughts on this? I'm less likely to try out experimental changes if the author has to invent a custom way to distribute them every time. It would be nice to have a known way that we use.

On my side I'm not using Quicklisp at all but rather packaging kons-9 up using Nix for my own uses. This works fine but it creates friction every time there is a change in the way kons-9 needs to be built e.g. the "asdf [monkey patching](https://en.wikipedia.org/wiki/Monkey_patch)" in krma .asd files to make the submodules override quicklisp.

EDIT: Let me reference [cl-bodge](https://github.com/borodust/cl-bodge#quicklisp) as an example of a project that is distributed via its own tailored Quicklisp distribution.

EDIT: Added wikipedia link to "monkey patching" since it is not Lisp lingo and may have caused accidental offense.

**Comment by awolven** [comment]
My advice to you is to not try out experimental code.

Using experimental code requires that you do the following:

1) know how git works
2) know how asdf works
3) know how quicklisp works
4) read the documentation/notes provided with the experimental distribution
5) have the ability to improvise at any juncture

Also, every three weeks someone opens a new issue for kons 9 saying we need
to work on a dependency management system.  Between git, asdf, and
quicklisp 99.9% of our problems are solved, and for the kons-9/main branch
you can usually get it installed in a handful of simple commands, which can
be provided in the readme.

cd ~/quicklisp/local-projects/
git clone https://github.com/kaveh808/kons-9.git
cd kons-9
git submodule update --init --recursive
sbcl
(ql:quickload :kons-9)
(kons-9:run)

**Reply by lukego** [reply]
> every three weeks someone opens a new issue for kons 9 saying we need
to work on a dependency management system

This sounds like a good problem to have: new developers are joining the project at a regular cadence and calibrating their expectations.

Hopefully having this labelled thread under *Discussions* will at least help new developers understand what has been discussed and proposed and rejected.

**Comment by lukego** [comment]
My $0.02 is that users-vs-developers is a false dichotomy. It's a spectrum and most people will fall around the middle. We ourselves are our own users.

If we want to create a good experience for "users" in the future then the first step is to create a good experience for "developers" in the present.

**Comment by bigos** [comment]
It's all talking without the list of problems to solve grouped by the OS, library, or another random environment factor. We do not have defined goals and we do not know how far we are from achieving our goals. How do we progress from the current state of discussion toward something that will help us gradually move toward our goals?

**Comment by awolven** [comment]
You say now in a later email that developers are using discussions to
calibrate their expectations, but what I actually saw is someone using
discussions, after having freely volunteered to help, to create a venue to
complain about how hard it was and refer to other developers as monkeys.

**Reply by lukego** [reply]
I'm sorry that it came across this way. That's not my intention at all. Discussion forums can be hard. I'll withdraw from this discussion now to avoid distracting you from the excellent work that you are doing.

**Reply by bigos** [reply]
I thought the complaint was aimed at me. Should I withdraw from the discussion as well?

**Comment by kaveh808** [comment]
Hey everyone,

Bear with me as it's 7am here and I'm still groggy from sleep. :)

Internet discussions are difficult. So much context is missing in text. I find it tricky at times, and have to remind myself not to take things personally.

I started this project -- my first open-source effort -- to build a community of people working towards a common goal. I have been encouraged by the participation so far. It has been a congenial experience to date and I would like to keep it so. And I do not want anyone to feel unwelcome and feel as if they have to withdraw from our discussions.

I don't think the term "monkey patching" was meant as an insult to anyone. It was a link to Wikipedia, where it's defined as a technical term derived from "guerrilla patching". Hacker terminology can be colorful and borderline insensitive. :)

I appreciate that everyone here is donating their time and skill because they care enough about the project to do so.

I'll address the specific questions of dependencies in a separate posting.

Regards,

Kaveh

**Reply by kaveh808** [reply]
Everyone is welcome to participate in the project, though it doesn't mean we will necessarily do everything you say. :)

And I want to foster an environment that caters to people feeling this way. I don't want to see any "votes of no confidence."

It is good to remind ourselves that we are all here because we want the project to succeed.

**Reply by bigos** [reply]
From my standpoint, it is not about unwelcome. Sometimes I feel that some people prefer me to stay in the peloton instead of racing against the stars. Maybe the cycling metaphor is not the best. And I do not try to compete. But sometimes, it is easy to be misunderstood.

**Reply by kaveh808** [reply]
It is very easy to be misunderstood online. I always try to assume the best of intentions on the writer's part, which why I emphasize that we are all here to work towards a common goal.

**Reply by foretspaisibles** [reply]
Always try to assume best intention on the writer's part is a great tip. Sleeping one night over messages we could answer with bad feelings is also working well. Taking time to reread a conversation with fresh eyes and paying more attention to the context are also good advice.

**Comment by kaveh808** [comment]
Regarding dependency management: this is a tough issue for a couple of reasons:

1) kons-9 is a platform for development: our users are developers. So I don't make a distinction between the two. I want it to be as simple as possible for our users/developers to get up and running.

2) I have a personal aversion to complex setup procedures. I admit it's an irrational feeling, but I feel confused and frustrated when getting some software up and running involves many steps that I don't understand. That's why I have been wanting to keep things simple for our users, as well as for my own selfish reasons.

Unfortunately I don't have much experience in this matter. Fortunately we have collaborators who do and can suggest approaches to consider. And I welcome those suggestions.

My gut feeling in designing software is to take the simplest route until there is a strong need to change course. For example, I started the development of the system using OpenGL 1.0 calls, even though they are by no means state of the art, because they were quick to get working. And it got the code up and running and has managed to get us this far. Now that there is a strong need to move to Vulkan, we are doing so.

Luke has suggested a number of approaches. I will admit that my only experience in dealing with dependencies has been to embed the font library code and ttf data directly into our source, to get the text engine working. Perhaps not terribly elegant, but it got the job done and kept installation simple. Well, as simple as CL installation gets. :)

As we expand our code base, for example with Andrew's new Vulkan code, I agree there is a need for some other approach, as we have to deal with external libraries such as Vulkan and Metal, compiling shaders, and the like.

So let's keep this discussion going. I'm confident we have the experience and talent to find a solution. :)

**Reply by bigos** [reply]
> So let's keep this discussion going. I'm confident we have the experience and talent to find a solution. :)

But do we have the patience?

**Comment by awolven** [comment]
We can develop an installer, which will have emacs, sbcl, slime, kons-9 and
all its dependencies for people who want the simple route.  For people who
already have a common lisp development environment, we can provide the
normal build instructions.

I understand the resistance to dividing consumers into "users" and
"developers" because of the seamless transition provided by common lisp.
But there still are newbies and hackers, or rather just people who don't
want a fuss and people who will tolerate some fuss to have a custom
environment, so I don't think it's a good idea to rely on a "one size fits
all" installation method.

**Comment by kaveh808** [comment]
Portable looks interesting, based on the docs. I do wonder if we can build on that, later down the road.

I agree that hackers should be able to customize their development environment. But I do want there to always be an "easy" installation setup to encourage those who want to give the system a try. As mentioned previously, I consider myself as someone yearning for an easy setup.

Let's wait for the step-by-step installation instructions that the team is working on for the existing system, and then see if we can keep at least the same complexity/dependencies while introducing the new graphics engine.

**Comment by ghost** [comment]


**Reply by bigos** [reply]
We are discussing various ideas. But how do we document the progress of the discussion so that we do not go in circles? The documentation standards that were proposed seem to apply to a finished idea. How do we draft the progress of a more complex discussion where we do not know the correct answer?  How we can handle the situation where we have several competing answers with different levels of verification of their usability? Does anybody have a suggestion?

**Comment by foretspaisibles** [comment]
At the risk of adding another dimension to the conversation, I feel that the very fact that this conversation happens is a sign the CL community is lacking a good baseline practice here. Generally speaking distributing applications on Linux is terribly hard which is why every programming language community is developing their own solution to that, so there is nothing really specific to CL here. If we come to a conclusive part it would be a great addition to the CL Cookbook.

I propose several actions and invite you to vote for or against them (arguments or references to the conversation so far are also welcome!). Please vote with the thumb up/down gitHub gadget.

Action 1, 2, 3 below.

**Comment by foretspaisibles** [comment]
Action 1 — Create a `kons-9-world` repository with `kons-9` and every source code dependency `kons-9` has as a git submodule. Interested collaborators can start experiment with that and figure out if that makes a simple and nice developer's experience while other collaborators do not need to change anything in their setup.

If the experiment is conclusive we make it the preferred way to work with kons-9, otherwise we just delete the repo and try something else.

PS: By “source code dependency” I mean a software dependency which is a source code whose latest version we need.

**Comment by foretspaisibles** [comment]
Action 2 — Bubble up that conversation to the CL Cookbook community. The CL cookbook should have easy guidelines about distributing software written in CL, so I feel confident it's worth checking with them if they are interested too.

**Comment by foretspaisibles** [comment]
Action 3 — Try to recruit other voices to that distribution guidelines discussions. There is several groups writing great CL software, and they have very valuable experience there we should take advantage of.

**Comment by awolven** [comment]
On "Action 1" what do you mean by "source" dependency?

**Reply by foretspaisibles** [reply]
By “source (code) dependency” I mean a software dependency which is a source code whose latest version we need. I added that comment to my “Action 1” card. (edit)

**Comment by awolven** [comment]
Going without a response to my question about Action #1, I'll comment
anyway:

I'm fine with using quicklisp to manage dependent systems which either:
1. never change
2. rarely/never have an API change

We just need to keep our finger on the pulse of the quicklisp repository we
are using.  (preferably the default)

The only time we really need to add the extra maintenance burden of a
submodule is when a specific version (i.e. commit hash) of the system needs
to be used in the kons-9 code, or by another submodule.  Or if you just
want to be extra safe about potential changes to that system.

-AKW

**Reply by bigos** [reply]
https://github.com/bigos/quicklisp-xref This is an unfinished project. I have run out of steam because I lacked the skills. It was a response to an idea of qucklisp maintainer. Should anybody write something like that? Is it possible to have a more humble version that only handles our quicklisp dependencies?

**Comment by awolven** [comment]
I would just clarify that we are talking about a "specific version" and not
necessarily a "latest version" for a git submodule.

**Comment by foretspaisibles** [comment]
I added a “not sure if I care (confused emoji)” button on my three little voting cards, so that people have a third choice. :-)

**Comment by kaveh808** [comment]
I have not voted on the 3 choices yet, as I want to have a better understanding of the work required by users to get kons-9 running.

Between git submodules and Quicklisp, what are the steps required to run kons-9 with the new Vulkan engine? Assuming the user is a CL users and knows how to use asdf and such.

Currently, it should be one step git checkout of the code to run the system. I'd like to keep it as a one step process.

**Comment by awolven** [comment]
The steps are:

git clone ***@***.***:awolven/kons-9.git
cd kons-9
git checkout --track -b krma origin/krma
git submodule update --init --recursive
sbcl
(ql:quickload :kons-9)
(kons-9:run)

**Comment by awolven** [comment]
The steps for kons-9 currently are:

git clone https://github.com/kaveh808/kons-9
cd kons-9
sbcl
(ql:quickload :kons-9)
(kons-9:run)

**Comment by awolven** [comment]
The steps to get kons-9 running after the merge:

git clone https://github.com/kaveh808/kons-9
cd kons-9
git submodule update --init --recursive
sbcl
(ql:quickload :kons-9)
(kons-9:run)
(ql:qui

**Comment by awolven** [comment]
For macos, there will be a one-time necessity of installing the vulkan sdk,
to get the MoltenVK driver and updating a file to make sure the path is
correct.

The sdk is redistributable, so we could avoid that step by using "git
submodule update --init --recursive".  And if you really wanted to, we
could publish bash and cmd scripts in the root directory of the kons-9 main
branch which could handle any branch switching and submodule updates.  So
two or three lines of git commands could be replaced by one line.

Otherwise you need an installer.  Developers can handle the normal setup.
Users are the problem, so providing an installer that sets up emacs/kons-9
executable (with quicklisp and asdf)/support files (i.e. MoltenVK driver).
Portacle could be used for this.

Also, in case you haven't been reading, Jacek and Luke and Marengo have not
complained about the "git submodule update --init --recursive" the topic is
squarely on some kind of modification or replacement of quicklisp, which is
completely unnecessary and only creates extra maintenance work and
complicates install for actual developers.

In engineering the first rule of problem solving is to clearly and
concisely describe the problem.  They should consider doing that before
promoting solutions.  Subjective adjectives like "hard" should be replaced
with concrete objective descriptions.

**Reply by bigos** [reply]
I need to add the last paragraph to the list of quotes I need to remember. 

With regards to clearly describing the problem, I have proposed a temporary solution. Looking at the quicklisp history of issues I have realized it is an unfinished project. While it works for many purposes it may not be the suitable foundation for us. 

There was some problem that took much of my attention and I have reduced my participation with the project. I do not want the frustrations from elsewhere affect the tone of the messages I post here. I will keep following the project, testing the progress and go with what you suggest. After all we can only test git submodules and see if it works for us. That will give more concrete answers to the problem we are trying to solve.

**Comment by awolven** [comment]
You may be wasting your time anyway, someone is already working on a
replacement to quicklisp which handles versions and such and all the
lacking things.  I think it was Eric Timmons, but I can't remember.  It's
probably best if you coordinate with someone who has already started and
can clearly define the problem.

**Comment by istewart-jinzoningen** [comment]
Hi everyone, just wanted to chime in as I've been watching the project from afar since inception, but just pulled down the code and got it running tonight. I'm on Linux and my only bump in the road was needing GLFW as a native-OS-level dependency; I think that and other dependencies should at least be noted and linked in the repository readme. 

From the discussion here, I can respect that a "one-stop-shop" install is the ideal, but having dependencies laid out on the front page of the repository is a worthy interim approach, and definitely the sort of thing I'm used to when encountering experimental and early-development open-source projects. On the other hand, I'm specifically using Gentoo, so I'm the kind of user who's not afraid of tracking down libraries and cobbling things together!

For any other Gentoo or rolling-release distro heads who may be having a look in, I did have to unmask the latest 3.3.8 version of GLFW from the testing side of Gentoo's repository. The 3.3.5 release tagged as stable seemed pretty far out of date, and I didn't want that to present an obstacle.

I also have an M1 MacBook Air handy, but I haven't updated Homebrew since I installed it shortly after getting the machine. If I get that up to date, I will try bringing kons-9 up on there and give some feedback.

**Reply by foretspaisibles** [reply]
Thank you for sharing your story, it's both interesting and helpful. Maybe it could be interesting if you share how you would detect if GLFW is available or not on your platform, that would allow the setup to redirect the user with missing requirements to the appropriate documentation with a useful diagnostic.

Just out of curiosity, if you have a few spare minutes could you put the output of

~~~
development/describe
~~~

in the discussion

https://github.com/kaveh808/kons-9/discussions/103

Have fun with KONS-9!

**Reply by istewart-jinzoningen** [reply]
Hey, sorry I dropped off from this discussion. Attempting to do 

> (ql:quickload "kons-9")

within SBCL resulted in the quickloading process being unable to find libglfw.so through the FFI. Unfortunately I didn't save a log of the error output, sorry. I had to Google what libglfw even was, and then on Gentoo it is a simple 

> emerge media-libs/glfw

I will add my system info to the other discussion.

**Reply by foretspaisibles** [reply]
Great, thanks! I'm happy you could go through the install, have fun with it! :-)
